File: ios/ExpoIombLibraryModule.swift
import ExpoModulesCore
import IOMbLibrary

// Define payload types for better type safety
struct SessionConfigurationPayload: Record {
  @Field var baseURL: String
  @Field var offerIdentifier: String
  @Field var hybridIdentifier: String?
}

struct LogViewEventPayload: Record {
  @Field var type: String
  @Field var category: String
  @Field var comment: String?
}

public class ExpoIombLibraryModule: Module {
  // helper functions
  private func stringToEnumValue(_ value: String, values: [String]) -> UInt {
    return UInt(values.firstIndex(where: { $0 == value }) ?? 0)
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoIombLibrary")

    AsyncFunction("sessionConfiguration") { (payload: SessionConfigurationPayload) -> Void in
      guard let url = URL(string: payload.baseURL) else { return }

      let configuration = IOMBSessionConfiguration(
        offerIdentifier: payload.offerIdentifier,
        hybridIdentifier: payload.hybridIdentifier,
        baseURL: url
      )

      IOMBSession.defaultSession(for: .iomb).start(with: configuration)
    }

    AsyncFunction("terminateSession") { () -> Void in
      IOMBSession.defaultSession(for: .iomb).terminateSession()
    }

    AsyncFunction("logViewEvent") { [self] (payload: LogViewEventPayload) -> Void in
      let types = ["appeared", "refreshed", "disappeared"]
      let eventType = IOMBViewEventType(rawValue: stringToEnumValue(payload.type, values: types))!
      let event = IOMBViewEvent(type: eventType, category: payload.category, comment: payload.comment)
      IOMBSession.defaultSession(for: .iomb).logEvent(event)
    }

    AsyncFunction("setDebugLogLevel") { [self] (_ level: String) -> Void in
      let levels = ["off", "error", "warning", "info", "trace"]
      let level = IOMBDebugLevel(rawValue: Int(stringToEnumValue(level, values: levels)))!
      IOMBLogging.setDebugLogLevel(level)
    }
  }
}

File: android/src/main/java/expo/modules/iomblibrary/ExpoIombLibraryModule.kt
package expo.modules.iomblibrary

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.records.Field

import de.infonline.lib.iomb.IOLViewEvent
import de.infonline.lib.iomb.IOMB
import de.infonline.lib.iomb.measurements.Measurement
import de.infonline.lib.iomb.measurements.iomb.IOMBSetup
import de.infonline.lib.iomb.IOLDebug

class SessionConfigurationPayload : Record {
  @Field
  val baseURL: String = ""
  @Field
  val offerIdentifier: String = ""
  @Field
  val hybridIdentifier: String? = null
}

class LogViewEventPayload : Record {
  @Field
  val type: String = ""
  @Field
  val category: String = ""
  @Field
  val comment: String? = null
}

class ExpoIombLibraryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoIombLibrary")

    AsyncFunction("sessionConfiguration") { payload: SessionConfigurationPayload ->
      val setup = IOMBSetup(
      offerIdentifier = payload.offerIdentifier,
      baseUrl = payload.baseURL,
      hybridIdentifier = payload.hybridIdentifier)

      IOMB_SESSION = IOMB.createBlocking(setup)
    }

    AsyncFunction("terminateSession") {
      IOMB.delete(Measurement.Type.IOMB)
    }

    AsyncFunction("logViewEvent") { payload: LogViewEventPayload ->
      val enumType = IOLViewEvent.IOLViewEventType.valueOf(uppercase(payload.type))
      val event = IOLViewEvent(type = enumType, category = payload.category, comment = payload.comment)
      IOMB_SESSION.logEvent(event)
    }

    AsyncFunction("setDebugModeEnabled") { enable: Boolean ->
      IOLDebug.debugMode = enable
    }
  }

  // Helper functions
  companion object {
    lateinit var IOMB_SESSION: Measurement
  }

  private fun uppercase(value: String) : String {
    return value.replaceFirstChar { it.uppercase() }
  }
}